File: ios/Solace/ViewModels/HomeViewModel.swift
import SwiftUI
import FirebaseFirestore

@MainActor
public final class HomeViewModel: ObservableObject {

    @Published public private(set) var uiConfig: SolaceUiConfig = .default
    @Published public private(set) var profiles: [UserProfile] = []
    @Published public private(set) var isLoading: Bool = true
    @Published public private(set) var error: String? = nil

    private let userRepo = UserRepository.shared
    private let configRepo = RemoteConfigRepository.shared
    private var profilesListener: ListenerRegistration?

    public init() {}

    deinit { profilesListener?.remove() }

    // ── Config ────────────────────────────────────────────────────────────

    public func loadConfig() async {
        uiConfig = await configRepo.fetchUiConfig()
    }

    // ── Profiles ──────────────────────────────────────────────────────────

    public func startObservingProfiles() {
        profilesListener?.remove()
        profilesListener = userRepo.observeProfiles { [weak self] profiles in
            Task { @MainActor in
                self?.profiles = profiles
                self?.isLoading = false
            }
        }
    }

    public func stopObservingProfiles() {
        profilesListener?.remove()
        profilesListener = nil
    }

    // ── Actions ───────────────────────────────────────────────────────────

    public func sendLike(from currentUid: String, to targetUid: String) {
        Task {
            do { try await userRepo.sendLike(from: currentUid, to: targetUid) }
            catch { self.error = error.localizedDescription }
        }
    }

    public func dismissError() { error = nil }
}

File: android/app/src/main/java/com/solace/ui/home/HomeViewModel.kt
package com.solace.ui.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.solace.data.config.RemoteConfigRepository
import com.solace.data.user.UserRepository
import com.solace.ui.config.HomeScreenConfig
import com.solace.ui.config.SolaceUiConfig
import com.solace.ui.config.UserProfile
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

data class HomeUiState(
    val uiConfig: SolaceUiConfig = SolaceUiConfig(),
    val profiles: List<UserProfile> = emptyList(),
    val isLoading: Boolean = true,
    val error: String? = null,
)

@HiltViewModel
class HomeViewModel @Inject constructor(
    private val userRepo: UserRepository,
    private val configRepo: RemoteConfigRepository,
) : ViewModel() {

    private val _uiState = MutableStateFlow(HomeUiState())
    val uiState: StateFlow<HomeUiState> = _uiState.asStateFlow()

    init {
        loadConfig()
        observeProfiles()
    }

    private fun loadConfig() {
        viewModelScope.launch {
            val config = configRepo.fetchUiConfig()
            _uiState.update { it.copy(uiConfig = config) }
        }
    }

    private fun observeProfiles() {
        viewModelScope.launch {
            userRepo.getProfiles()
                .catch { e -> _uiState.update { it.copy(error = e.message, isLoading = false) } }
                .collect { profiles ->
                    _uiState.update { it.copy(profiles = profiles, isLoading = false) }
                }
        }
    }

    fun sendLike(currentUid: String, targetUid: String) {
        viewModelScope.launch {
            userRepo.sendLike(fromUid = currentUid, toUid = targetUid)
        }
    }

    fun dismissError() = _uiState.update { it.copy(error = null) }
}
